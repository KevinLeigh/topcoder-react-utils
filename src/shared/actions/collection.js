/**
 * @module "actions.collection"
 * @desc Actions for management of *Collection* segment of Redux store.
 */

import _ from 'lodash';
import * as redux from 'utils/redux';

/**
 * @static
 * @desc Creates an action that adds items to the collection.
 * @param {Object} index The index of items to add. It is a JS object with
 *  ID-to-item mapping. Each item will be added to the collection under the
 *  specified index.
 * @return {Action}
 */
function addItems(index) {
  return { index, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that increments reference counters of the specified
 *  items.
 * @param {String|String[]} ids Item ID, or an array of IDs.
 * @param {Number} factor=1 Optional. The increment value.
 * @return {Action}
 */
function bookItems(ids, factor = 1) {
  return { factor, ids: _.isArray(ids) ? ids : [ids] };
}

/**
 * @static
 * @desc Creates an action that removes from the collection all items that
 *  (i) are not used (i.e. their reference counters equal to zero); (ii) were
 *  added to the collection before the specified `timestamp`.
 * @param {Number} timestamp=Number.MAX_VALUE The minimal timestamp value of
 *  items that should be kept in the collection even if they are not currently
 *  used.
 * @return {Action}
 */
function clean(timestamp = Number.MAX_VALUE) {
  return timestamp;
}

/**
 * @static
 * @desc Creates an action that decrements reference counters of the specified
 *  collection items.
 * @param {String|String[]} ids Item ID, or an array of IDs.
 * @param {Number} factor=1 Optional. Decrement value.
 * @return {Action}
 */
function freeItems(ids, factor = 1) {
  return { factor, ids: _.isArray(ids) ? ids : [ids] };
}

/**
 * @static
 * @desc Creates an action that signals beginning of an async loading of the
 *  specified item.
 * @param {String} operationId Operation UUID, to match with the corresponding
 *  {@link module:actions.collection.loadItemDone} action.
 * @param {String} itemId Item ID.
 * @return {Action}
 */
function loadItemInit(operationId, itemId) {
  return { itemId, operationId };
}

/**
 * @static
 * @desc Creates an action that finalizes an async loading of the specified
 *  item.
 * @param {String} operationId Operation UUID, to match with the corresponding
 *  {@link module:actions.collection.loadItemDone} action.
 * @param {String} itemId Item ID.
 * @param {Any} item The loaded item, to be added into the collection.
 * @return {Action}
 */
function loadItemDone(operationId, itemId, item) {
  return {
    item,
    itemId,
    operationId,
    timestamp: Date.now(),
  };
}

export default redux.createActions({
  COLLECTION: {
    ADD_ITEMS: addItems,
    BOOK_ITEMS: bookItems,
    CLEAN: clean,
    FREE_ITEMS: freeItems,
    LOAD_ITEM_INIT: loadItemInit,
    LOAD_ITEM_DONE: loadItemDone,
  },
});
