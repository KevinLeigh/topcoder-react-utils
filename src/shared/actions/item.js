/**
 * @module "actions.item"
 * @desc Actions for management of item data in Redux store.
 */

import * as redux from 'utils/redux';

/**
 * @static
 * @desc Creates an action that drops item's data if (i) the value of item's
 *  reference counter is not positive; (ii) item's timestamp is older than
 *  `olderThan` value.
 * @param {Number} olderThan=Number.MAX_VALUE Optional. Timestamp [ms].
 *  If provided, item's  data will be dropped only if they were received / set
 *  prior to this moment in time.
 * @return {Action}
 */
function dropData(olderThan = Number.MAX_VALUE) {
  return olderThan;
}

/**
 * @static
 * @desc Creates an action that signals the beginning of async loading of
 *  item's data.
 * @param {Number|String} loadingOperationId The unique identifier of the
 *  loading operation. It will be written into item's segment of Redux store
 *  with the following efffects:
 *
 *  - We consider that item's data are being loaded if, and only if,
 *    the item's `loadingOperationId` has a truly value;
 *
 *  - `loadingOperationId` is used to match the corresponding
 *    {@link module:actions.item.loadDataDone} operation.
 *
 * @return {Action}
 */
function loadDataInit(loadingOperationId) {
  return loadingOperationId;
}

/**
 * @static
 * @desc Creates an action that handles the result of async loading of item's
 *  data.
 * @param {Number|String} loadingOperationId The unique identifier of the
 *  loading operation. If it does not match the value currently written for
 *  the item in the Redux store (see {@link module:actions.item.loadDataInit}),
 *  the action will be silently ignored.
 * @param {Any} data Any value to store as the item's data.
 * @return {Action}
 */
function loadDataDone(loadingOperationId, data) {
  return { data, loadingOperationId, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that stores given data into the item.
 * @param {Any} data Any value to store as the item's data.
 * @return {Action}
 */
function setData(data) {
  return { data, timestamp: Date.now() };
}

/**
 * @static
 * @desc Creates an action that updates the item's reference counter.
 * @param {Number} shift The value to add to the item's reference counter.
 *  Should be `1` or `-1` in majority of cases, although any other values
 *  are permitted.
 * @return {Action}
 */
function updateReferenceCounter(shift) {
  return shift;
}

export default redux.createActions({
  ITEM: {
    DROP_DATA: dropData,
    LOAD_DATA_INIT: loadDataInit,
    LOAD_DATA_DONE: loadDataDone,
    SET_DATA: setData,
    UPDATE_REFERENCE_COUNTER: updateReferenceCounter,
  },
});
