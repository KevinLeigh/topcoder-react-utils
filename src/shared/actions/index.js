import collection from './collection';
import item from './item';

export default {
  ...collection,
  ...item,
};
