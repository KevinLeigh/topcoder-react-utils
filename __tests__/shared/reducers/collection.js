/**
 * Testing of collection reducer.
 */

import actions from 'actions/collection';
import reducer from 'reducers/collection';
import MockDate from 'mockdate';

afterAll(() => MockDate.reset());

describe('Generic Snapshot Testing', () => {
  /* In this group of tests we just dispatch a bunch of collection actions that
   * initialize and modify the same state object. After each modification, make
   * a snapshot of the state, and compare it to the expected one. */

  const a = actions.collection;
  let state;

  test('State initialization', () => {
    MockDate.set(1);
    state = reducer(undefined, '@@INIT');
    expect(state).toMatchSnapshot();
  });

  test('Addition of items and booking', () => {
    MockDate.set(2);
    state = reducer(state, a.addItems({ id1: 'Item1', id2: 'Item2' }));
    expect(state).toMatchSnapshot();
    MockDate.set(3);
    state = reducer(state, a.bookItems('id3'));
    expect(state).toMatchSnapshot();
    MockDate.set(4);
    state = reducer(state, a.addItems({ id3: 'Item3' }));
    expect(state).toMatchSnapshot();
    MockDate.set(5);
    state = reducer(state, a.bookItems(['id1', 'id2', 'id3'], 2));
    expect(state).toMatchSnapshot();
  });

  test('Loading of items', () => {
    MockDate.set(6);
    state = reducer(state, a.loadItemInit('op1', 'id4'));
    expect(state).toMatchSnapshot();
    MockDate.set(7);
    state = reducer(state, a.bookItems('id5'));
    MockDate.set(8);
    state = reducer(state, a.loadItemInit('op2', 'id5'));
    expect(state).toMatchSnapshot();
    MockDate.set(9);
    state = reducer(state, a.loadItemDone('op1', 'id4', 'Item4'));
    expect(state).toMatchSnapshot();
    MockDate.set(10);
    state = reducer(state, a.loadItemInit('op3', 'id5'));
    expect(state).toMatchSnapshot();
    MockDate.set(11);
    state = reducer(state, a.loadItemDone('op2', 'id5', 'Item5a'));
    expect(state).toMatchSnapshot();
    MockDate.set(12);
    state = reducer(state, a.loadItemDone('op3', 'id5', 'Item5b'));
    expect(state).toMatchSnapshot();
  });
});
