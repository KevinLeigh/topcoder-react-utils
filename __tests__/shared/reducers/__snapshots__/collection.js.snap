// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic Snapshot Testing Addition of items and booking 1`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 2,
  },
}
`;

exports[`Generic Snapshot Testing Addition of items and booking 2`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 2,
  },
  "id3": Object {
    "item": null,
    "loadingOperationId": "",
    "numRefs": 1,
    "timestamp": 0,
  },
}
`;

exports[`Generic Snapshot Testing Addition of items and booking 3`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 1,
    "timestamp": 4,
  },
}
`;

exports[`Generic Snapshot Testing Addition of items and booking 4`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
}
`;

exports[`Generic Snapshot Testing Loading of items 1`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
  "id4": Object {
    "item": null,
    "loadingOperationId": "op1",
    "numRefs": 0,
    "timestamp": 0,
  },
}
`;

exports[`Generic Snapshot Testing Loading of items 2`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
  "id4": Object {
    "item": null,
    "loadingOperationId": "op1",
    "numRefs": 0,
    "timestamp": 0,
  },
  "id5": Object {
    "item": null,
    "loadingOperationId": "op2",
    "numRefs": 1,
    "timestamp": 0,
  },
}
`;

exports[`Generic Snapshot Testing Loading of items 3`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
  "id4": Object {
    "item": "Item4",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 9,
  },
  "id5": Object {
    "item": null,
    "loadingOperationId": "op2",
    "numRefs": 1,
    "timestamp": 0,
  },
}
`;

exports[`Generic Snapshot Testing Loading of items 4`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
  "id4": Object {
    "item": "Item4",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 9,
  },
  "id5": Object {
    "item": null,
    "loadingOperationId": "op3",
    "numRefs": 1,
    "timestamp": 0,
  },
}
`;

exports[`Generic Snapshot Testing Loading of items 5`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
  "id4": Object {
    "item": "Item4",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 9,
  },
  "id5": Object {
    "item": null,
    "loadingOperationId": "op3",
    "numRefs": 1,
    "timestamp": 0,
  },
}
`;

exports[`Generic Snapshot Testing Loading of items 6`] = `
Object {
  "id1": Object {
    "item": "Item1",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id2": Object {
    "item": "Item2",
    "loadingOperationId": "",
    "numRefs": 2,
    "timestamp": 2,
  },
  "id3": Object {
    "item": "Item3",
    "loadingOperationId": "",
    "numRefs": 3,
    "timestamp": 4,
  },
  "id4": Object {
    "item": "Item4",
    "loadingOperationId": "",
    "numRefs": 0,
    "timestamp": 9,
  },
  "id5": Object {
    "item": "Item5b",
    "loadingOperationId": "",
    "numRefs": 1,
    "timestamp": 12,
  },
}
`;

exports[`Generic Snapshot Testing State initialization 1`] = `Object {}`;
